diff --git a/test cases/cmake/26 target file/main.cpp b/test cases/cmake/26 target file/main.cpp
new file mode 100644
index 000000000..250d41e1a
--- /dev/null
+++ b/test cases/cmake/26 target file/main.cpp	
@@ -0,0 +1,6 @@
+#include "final.hpp"
+#include <iostream>
+
+using namespace std;
+
+int main(void) { testGetStr(); }
diff --git a/test cases/cmake/26 target file/meson.build b/test cases/cmake/26 target file/meson.build
new file mode 100644
index 000000000..cc9cc0c34
--- /dev/null
+++ b/test cases/cmake/26 target file/meson.build	
@@ -0,0 +1,12 @@
+project('cmake_code_gen', ['c', 'cpp'])
+
+cm = import('cmake')
+
+# Subproject with the "code generator"
+sub_pro = cm.subproject('cmMod')
+sub_final = sub_pro.dependency('final', include_type: 'system')
+
+# Build the exe
+exe = executable('main', ['main.cpp'], dependencies: [sub_final])
+
+test('test', exe)
diff --git a/test cases/cmake/26 target file/subprojects/cmMod/CMakeLists.txt b/test cases/cmake/26 target file/subprojects/cmMod/CMakeLists.txt
new file mode 100644
index 000000000..efb0046fd
--- /dev/null
+++ b/test cases/cmake/26 target file/subprojects/cmMod/CMakeLists.txt	
@@ -0,0 +1,31 @@
+cmake_minimum_required(VERSION 3.7)
+
+project(cmMod)
+set(CMAKE_CXX_STANDARD 14)
+
+add_executable(cm_exe_local local.cpp)
+
+add_executable(cm_exe_imported IMPORTED)
+set_property(TARGET cm_exe_imported PROPERTY IMPORTED_LOCATION "/usr/bin/cp")
+
+
+add_custom_command(OUTPUT ${CMAKE_CURRENT_BINARY_DIR}/output1.cpp
+  COMMAND $<TARGET_FILE:cm_exe_local> ${PROJECT_SOURCE_DIR}/input1.txt ${CMAKE_CURRENT_BINARY_DIR}/output1.cpp
+  DEPENDS cm_exe_local
+  DEPENDS input1.txt)
+add_custom_command(OUTPUT ${CMAKE_CURRENT_BINARY_DIR}/output2.cpp
+  COMMAND $<TARGET_FILE:cm_exe_imported> ${PROJECT_SOURCE_DIR}/input2.txt ${CMAKE_CURRENT_BINARY_DIR}/output2.cpp
+  DEPENDS cm_exe_imported
+  DEPENDS input2.txt)
+
+add_custom_target(output1_cpp
+  DEPENDS "${CMAKE_CURRENT_BINARY_DIR}/output1.cpp")
+add_custom_target(output2_cpp
+  DEPENDS "${CMAKE_CURRENT_BINARY_DIR}/output2.cpp")
+
+add_library(final SHARED final.cpp ${CMAKE_CURRENT_BINARY_DIR}/output1.cpp ${CMAKE_CURRENT_BINARY_DIR}/output2.cpp)
+add_dependencies(final output1_cpp output2_cpp)
+target_include_directories(final
+    PUBLIC
+        ${PROJECT_SOURCE_DIR}/include
+)
diff --git a/test cases/cmake/26 target file/subprojects/cmMod/final.cpp b/test cases/cmake/26 target file/subprojects/cmMod/final.cpp
new file mode 100644
index 000000000..07fe27540
--- /dev/null
+++ b/test cases/cmake/26 target file/subprojects/cmMod/final.cpp	
@@ -0,0 +1,9 @@
+#include <cassert>
+#include <string>
+
+std::string getStr1();
+std::string getStr2();
+void testGetStr() {
+  assert(getStr1() == "Hello World 1");
+  assert(getStr2() == "Hello World 2");
+}
diff --git a/test cases/cmake/26 target file/subprojects/cmMod/include/final.hpp b/test cases/cmake/26 target file/subprojects/cmMod/include/final.hpp
new file mode 100644
index 000000000..20f7335e5
--- /dev/null
+++ b/test cases/cmake/26 target file/subprojects/cmMod/include/final.hpp	
@@ -0,0 +1 @@
+void testGetStr();
diff --git a/test cases/cmake/26 target file/subprojects/cmMod/input1.txt b/test cases/cmake/26 target file/subprojects/cmMod/input1.txt
new file mode 100644
index 000000000..0bc73a7a2
--- /dev/null
+++ b/test cases/cmake/26 target file/subprojects/cmMod/input1.txt	
@@ -0,0 +1 @@
+example content
diff --git a/test cases/cmake/26 target file/subprojects/cmMod/input2.txt b/test cases/cmake/26 target file/subprojects/cmMod/input2.txt
new file mode 100644
index 000000000..7ddb6aaf9
--- /dev/null
+++ b/test cases/cmake/26 target file/subprojects/cmMod/input2.txt	
@@ -0,0 +1,4 @@
+#include <string>
+std::string getStr2() {
+  return "Hello World 2";
+}
diff --git a/test cases/cmake/26 target file/subprojects/cmMod/local.cpp b/test cases/cmake/26 target file/subprojects/cmMod/local.cpp
new file mode 100644
index 000000000..9bf213b88
--- /dev/null
+++ b/test cases/cmake/26 target file/subprojects/cmMod/local.cpp	
@@ -0,0 +1,30 @@
+#include <cassert>
+#include <fstream>
+#include <iostream>
+#include <sstream>
+
+using namespace std;
+
+int main(int argc, const char *argv[]) {
+  if (argc < 3) {
+    cerr << argv[0] << " requires an input and output file!" << endl;
+    return 1;
+  }
+
+  ifstream in(argv[1]);
+  cerr << in.is_open() << "\t" << argv[1];
+  stringstream buffer;
+  buffer << in.rdbuf();
+  cerr << buffer.str() << "\n";
+  assert(buffer.str() == "example content\n");
+
+  ofstream out(argv[2]);
+  out << R"(
+#include <string>
+std::string getStr1() {
+  return "Hello World 1";
+}
+)";
+
+  return 0;
+}
