diff --git a/mesonbuild/cmake/generator.py b/mesonbuild/cmake/generator.py
index a4224ccb2..6204029ec 100644
--- a/mesonbuild/cmake/generator.py
+++ b/mesonbuild/cmake/generator.py
@@ -95,7 +95,8 @@ def parse_generator_expressions(
             return ';'.join([x for x in tgt.properties[f'IMPORTED_LOCATION_{cfg}'] if x])
         elif 'IMPORTED_LOCATION' in tgt.properties:
             return ';'.join([x for x in tgt.properties['IMPORTED_LOCATION'] if x])
-        return ''
+        else:
+            return str(tgt.build_path)
 
     supported = {
         # Boolean functions
diff --git a/mesonbuild/cmake/traceparser.py b/mesonbuild/cmake/traceparser.py
index 04c40a9bd..9b03834c1 100644
--- a/mesonbuild/cmake/traceparser.py
+++ b/mesonbuild/cmake/traceparser.py
@@ -54,6 +54,7 @@ class CMakeTarget:
     def __init__(
                 self,
                 name:        str,
+                build_path:  T.Optional[Path],
                 target_type: str,
                 properties:  T.Optional[T.Dict[str, T.List[str]]] = None,
                 imported:    bool                                 = False,
@@ -62,6 +63,7 @@ class CMakeTarget:
         if properties is None:
             properties = {}
         self.name            = name
+        self.build_path      = build_path
         self.type            = target_type
         self.properties      = properties
         self.imported        = imported
@@ -87,7 +89,7 @@ class CMakeTarget:
 
 class CMakeGeneratorTarget(CMakeTarget):
     def __init__(self, name: str) -> None:
-        super().__init__(name, 'CUSTOM', {})
+        super().__init__(name, None, 'CUSTOM', {})
         self.outputs = []        # type: T.List[Path]
         self._outputs_str = []   # type: T.List[str]
         self.command = []        # type: T.List[T.List[str]]
@@ -108,6 +110,7 @@ class CMakeTraceParser:
         self.env = env
         self.permissive = permissive  # type: bool
         self.cmake_version = cmake_version  # type: str
+        self.build_dir = build_dir
         self.trace_file = 'cmake_trace.txt'
         self.trace_file_path = build_dir / self.trace_file
         self.trace_format = 'json-v1' if version_compare(cmake_version, '>=3.17') else 'human'
@@ -360,16 +363,16 @@ class CMakeTraceParser:
         args = list(tline.args) # Make a working copy
 
         # Make sure the exe is imported
-        is_imported = True
-        if 'IMPORTED' not in args:
-            return self._gen_exception('add_executable', 'non imported executables are not supported', tline)
-
-        args.remove('IMPORTED')
-
-        if len(args) < 1:
-            return self._gen_exception('add_executable', 'requires at least 1 argument', tline)
-
-        self.targets[args[0]] = CMakeTarget(args[0], 'EXECUTABLE', {}, tline=tline, imported=is_imported)
+        if 'IMPORTED' in args:
+            args.remove('IMPORTED')
+            if len(args) < 1:
+                return self._gen_exception('add_executable', 'requires at least 1 argument', tline)
+            self.targets[args[0]] = CMakeTarget(args[0], None, 'EXECUTABLE', {}, tline=tline, imported=True)
+        else:
+            if len(args) < 1:
+                return self._gen_exception('add_executable', 'requires at least 1 argument', tline)
+            build_path = self.build_dir.parent / args[0]
+            self.targets[args[0]] = CMakeTarget(args[0], build_path, 'EXECUTABLE', {}, tline=tline, imported=False)
 
     def _cmake_add_library(self, tline: CMakeTraceLine) -> None:
         # DOC: https://cmake.org/cmake/help/latest/command/add_library.html
@@ -382,7 +385,7 @@ class CMakeTraceParser:
             if len(args) < 1:
                 return self._gen_exception('add_library', 'interface library name not specified', tline)
 
-            self.targets[args[0]] = CMakeTarget(args[0], 'INTERFACE', {}, tline=tline, imported='IMPORTED' in args)
+            self.targets[args[0]] = CMakeTarget(args[0], None, 'INTERFACE', {}, tline=tline, imported='IMPORTED' in args)
         elif 'IMPORTED' in args:
             args.remove('IMPORTED')
 
@@ -390,7 +393,7 @@ class CMakeTraceParser:
             if len(args) < 2:
                 return self._gen_exception('add_library', 'requires at least 2 arguments', tline)
 
-            self.targets[args[0]] = CMakeTarget(args[0], args[1], {}, tline=tline, imported=True)
+            self.targets[args[0]] = CMakeTarget(args[0], None, args[1], {}, tline=tline, imported=True)
         elif 'ALIAS' in args:
             args.remove('ALIAS')
 
@@ -399,11 +402,11 @@ class CMakeTraceParser:
                 return self._gen_exception('add_library', 'requires at least 2 arguments', tline)
 
             # Simulate the ALIAS with INTERFACE_LINK_LIBRARIES
-            self.targets[args[0]] = CMakeTarget(args[0], 'ALIAS', {'INTERFACE_LINK_LIBRARIES': [args[1]]}, tline=tline)
+            self.targets[args[0]] = CMakeTarget(args[0], None, 'ALIAS', {'INTERFACE_LINK_LIBRARIES': [args[1]]}, tline=tline)
         elif 'OBJECT' in args:
             return self._gen_exception('add_library', 'OBJECT libraries are not supported', tline)
         else:
-            self.targets[args[0]] = CMakeTarget(args[0], 'NORMAL', {}, tline=tline)
+            self.targets[args[0]] = CMakeTarget(args[0], None, 'NORMAL', {}, tline=tline)
 
     def _cmake_add_custom_command(self, tline: CMakeTraceLine, name: T.Optional[str] = None) -> None:
         # DOC: https://cmake.org/cmake/help/latest/command/add_custom_command.html
